tools/merger.py

```diff
@@ -72,78 +72,68 @@ def load_merger(filenames, return_mass=False, trim=True):
+    Ntot = 0
+    dtypes = {}
+    
+    if do_prog:
+        N_prog_tot = 0
+        fields.remove('Progenitors')  # treat specially
+    
+    for fn in filenames:
+        with asdf.open(fn) as af:
+            # Peek at the first field to get the total length
+            # If the lengths of fields don't match up, that will be an error later
+            Ntot += len(af['data'][fields[0]])
+            
+            for field in fields:
+                if field not in dtypes:
+                    dtypes[field] = af['data'][field].dtype
+            
+            if do_prog:
+                N_prog_tot += len(af['data']['Progenitors'])
+                
+    # Make the empty tables
+    t = Table({f:np.empty(Ntot, dtype=dtypes[f]) for f in fields}, copy=False)
+    if do_prog:
+        p = Table({'Progenitors':np.empty(N_prog_tot, dtype=np.int64)}, copy=False)
+    
+    # Fill the data into the empty tables
+    j = 0
+    jp = 0
+    for i, fn in enumerate(filenames):
+        print(f"File number {i+1:d} of {len(filenames)}")
         f = asdf.open(fn, lazy_load=True, copy_arrays=True)
         fdata = f['data']
-
-        N_halos = fdata[fields[0]].shape[0]
-        num_progs_cum = 0
-        if i == 0:
-
-            dtypes = []
-            for field in fields:
-                dtype = fdata[field].dtype
-                try:
-                    shape = fdata[field].shape[1]
-                    dtype = (dtype, shape)
-                except:
-                    pass
-                if field != "Progenitors":
-                    dtypes.append((field, dtype))
-
-                if field == "Progenitors":
-                    final_progs = fdata["Progenitors"]
-                    try:
-                        num_progs = fdata["NumProgenitors"] + num_progs_cum
-                    except:
-                        print(
-                            "You need to also request 'NumProgenitors' if requesting the 'Progenitors' field"
-                        )
-                        exit()
-                    num_progs_cum += np.sum(num_progs)
-
-            final = np.empty(N_halos, dtype=dtypes)
-            for field in fields:
-                if "Progenitors" != field:
-                    final[field] = fdata[field]
-
-            if "Progenitors" in fields:
-                final["NumProgenitors"] = num_progs
-
-        else:
-            new = np.empty(N_halos, dtype=dtypes)
-            for field in fields:
-
-                if field != "Progenitors":
-                    new[field] = fdata[field]
-
-                if field == "Progenitors":
-                    progs = fdata["Progenitors"]
-                    final_progs = np.hstack((final_progs, progs))
-
-                    num_progs = fdata["NumProgenitors"] + num_progs_cum
-                    num_progs_cum += np.sum(num_progs)
-
-            if "Progenitors" in fields:
-                new["NumProgenitors"] = num_progs
-
-            final = np.hstack((final, new))
```

Comment:
This is a classic Numpy anti-pattern!  What's happening now is that each time you load a new file, all the previous data has to be copied too.  Numpy hstack, or any flavor of append, is never a good idea on large data (fine for small data, of course).  Instead, if the number of eventual elements to be read is known, count the number of elements, allocate one array of that size (per column), then read the data into that pre-allocated array.

Reply:
Also, probably don't want to pack data that comes out of ASDF into a Numpy structured array.  That packs all the columns together.  Keep the columns separate, e.g. by using an Astropy table.